[AARCH64] Add support for Cortex-A320

stuij opened the pull request:
This patch adds initial support for the recently announced Armv9 Cortex-A320 processor.

For more information, including the Technical Reference Manual, see: https://developer.arm.com/Processors/Cortex-A320

stuij committed:
[AARCH64] Add support for Cortex-A320

This patch adds initial support for the recently announced Armv9
Cortex-A320 processor.

For more information, including the Technical Reference Manual, see:
https://developer.arm.com/Processors/Cortex-A320

Co-authored-by: Oliver Stannard <oliver.stannard@arm.com>

Stylie777 commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -1109,10 +1125,12 @@ class ProcessorAlias<string n, string alias> {
-                     [TuneA35]>;
 def : ProcessorModel<"cortex-a34", CortexA53Model, ProcessorFeatures.A53,
                      [TuneA35]>;
+def : ProcessorModel<"cortex-a35", CortexA53Model, ProcessorFeatures.A53,
```
nit: What is the reason for repositioning `cortex-a35` here? Is it to follow an alphabetical order?

stuij replied:
Mostly to keep it in line with the rest of the places here in the code where we enumerate the cores. In my head canon the logic is 'subsequent iterations of the same class of cores'.

davemgreen reviewed (commented):
Can you add a release-note that this new CPU has been added?

jthackray commented on llvm/lib/TargetParser/Host.cpp:
```diff
@@ -228,6 +228,7 @@ StringRef sys::detail::getHostCPUNameForARM(StringRef ProcCpuinfoContent) {
         .Case("0xd14", "cortex-r82ae")
         .Case("0xd02", "cortex-a34")
         .Case("0xd04", "cortex-a35")
+        .Case("0xd8f", "cortex-a320")
```
There is also `llvm/unittests/TargetParser/Host.cpp` which reverse tests these, but it doesn't appear to be very well maintained (and I'm not sure how useful it is; perhaps that's why it's not been updated recently, and because there are so many cores now supported, it needs removing or some TLC)

stuij replied:
Added the test regardless.

stuij committed:
Address review comments

- add release note entry for Cortex-A320
- add AArch64CPUTestParams test
- add Linux host cpu name test

stuij commented on llvm/unittests/TargetParser/TargetParserTest.cpp:
```diff
@@ -1262,7 +1262,7 @@ INSTANTIATE_TEST_SUITE_P(
 
 // Note: number of CPUs includes aliases.
-static constexpr unsigned NumAArch64CPUArchs = 89;
+static constexpr unsigned NumAArch64CPUArchs = 90;
```
done on both release notes and AArch64CPUTestParams test.

jthackray reviewed (approved):
LGTM

Stylie777 reviewed (approved):
LGTM
